Framework/DataHandling/src/SaveAscii2.cpp

```diff
@@ -247,57 +249,88 @@ void SaveAscii2::exec() {
       }
     }
   }
-  if (!idx.empty()) {
-    nSpectra = static_cast<int>(idx.size());
+
+  // if no interval or spectra list, take all of them
+  if (idx.empty()) {
+    for (int i = 0; i < nSpectra; i++) {
+      idx.insert(i);
+    }
   }
 
   if (m_nBins == 0 || nSpectra == 0) {
     throw std::runtime_error("Trying to save an empty workspace");
   }
-  std::ofstream file(filename.c_str(), (appendToFile ? std::ios::app : std::ios::out));
 
-  if (!file) {
-    g_log.error("Unable to create file: " + filename);
-    throw Exception::FileError("Unable to create file: ", filename);
-  }
-  // Set the number precision
-  if (prec != EMPTY_INT()) {
-    file.precision(prec);
-  }
-  if (scientific) {
-    file << std::scientific;
-  }
-  const std::vector<std::string> logList = getProperty("LogList");
-  if (!logList.empty()) {
-    writeFileHeader(logList, file);
-  }
-  if (writeHeader) {
-    file << comment << " X " << m_sep << " Y " << m_sep << " E";
-    if (m_writeDX) {
-      file << " " << m_sep << " DX";
-    }
-    file << '\n';
-  }
+  bool OneSpectrumPerFile = getProperty("OneSpectrumPerFile");
+
+  Progress progress(this, 0.0, 1.0, idx.size());
+
   // populate the meta data map
   if (!m_metaData.empty()) {
     populateAllMetaData();
   }
-  if (idx.empty()) {
-    Progress progress(this, 0.0, 1.0, nSpectra);
-    for (int i = 0; i < nSpectra; i++) {
-      writeSpectrum(i, file);
-      progress.report();
+
+  auto idxIt = idx.begin();
+  while (idxIt != idx.end()) {
+    std::string currentFilename = filename;
+    size_t extPosition;
+    if (OneSpectrumPerFile) {
+      for (const std::string &ext : ASCII_EXTS) {
+        extPosition = filename.find(ext);
+        if (extPosition != std::string::npos)
+          break;
+      }
+      std::ostringstream ss;
+      double axisValue = m_ws->getAxis(1)->getValue(*idxIt);
+      std::string axisLabel = m_ws->getAxis(1)->unit()->label();
+      ss << std::scientific;
+      ss << std::string(filename, 0, extPosition) << "_" << *idxIt;
+      if (!(m_ws->getAxis(1)->isSpectra()))
+        ss << "_" << axisValue << axisLabel;
+      ss << std::string(filename, extPosition);
+      currentFilename = ss.str();
     }
-  } else {
-    Progress progress(this, 0.0, 1.0, idx.size());
-    for (int i : idx) {
-      writeSpectrum(i, file);
+    std::ofstream file(currentFilename, (appendToFile ? std::ios::app : std::ios::out));
+
+    if (file.bad()) {
+      g_log.error("Unable to create file: " + currentFilename);
+      throw Exception::FileError("Unable to create file: ", currentFilename);
```

Comment:
If one spectrum per file is asked, I think it's better not to throw, but rather skip the failed file with an error log. It would also be useful to add the actual reason of the failure to the message.

Reply:
As discussed, I kept the exception but removed the log as `throw` will lead to a log anyway.